multi-era/rust/src/utils.rs

```diff
@@ -233,6 +231,24 @@ impl MultiEraBlock {
             Self::Conway(block) => block.invalid_transactions.clone(),
         }
     }
+
+    pub fn hash(&self) -> BlockHeaderHash {
+         let bytes = match self {
+             Self::Byron(block) => match block {
+                 // why there's no to_bytes or sth for byron block headers?
```

Comment:
@gostkin  There should be. I think it's the `ToBytes` trait? We don't implement the same traits as the other eras since we can't do encoding preservation/etc so it directly uses cbor_event's, but byron/cip25 should still have that `serialization::ToBytes` (in `core/rust/src/serialization.rs`). There's a TODO about regeneration there so we might want to think about if we want to keep this long-term or if there was a way to use the other traits for non-preserve types. For Byron it's not a big deal since everything is canonical anyway but for cip25 it's weird since some things are subsets but maybe those aren't implementing any of the traits.

Reply:
More context: we had these `ToBytes`/`FromBytes` traits from before we did preserve-encodings. (possibly named differently e.g.`ToFromBytes` or `ToCBORBytes/FromCBORBytes` - I don't remember). The new CML's `Deserialize`  which now has those to/from bytes methods now has the ability to preserve encodings and/or force canonical encoding which byron/cip25 don't.

Byron could kind of fake it since it's all guaranteed to be canonical anyway so both methods would do the same thing. This would have to be changed in cddl-codegen though since when you're not preserving encodings in general we don't make any assumptions about all inputs being canonical so we don't implement these traits and instead directly implement `cbor_event::Serialize` like we did before. You couldn't blanket impl it like with `ToBytes` since it would blanket impl it for everything not just the byron types.

CIP25 is weird since it's not canonical but skips some data e.g. parses a subset of objects so round-tripping isn't preserved. We could maybe regen cip25 and have it impl those types with the exception of the outermost object (the one equivalent to a tx metadatum) which IIRC doesn't implement `ToBytes` anyway.